[feat] Allow multiple comments on same range

joassouza opened the pull request:
Com este PR a gente permite adicionar mais de um comentário em um mesmo trecho do texto. 
[Trello](https://trello.com/c/SUFzBIbC/172)

joassouza committed:
[feat] Allow add multiple comments on same range

joassouza committed:
[test] Add test to add comment on same range

joassouza committed:
[test] Add test to validate if a new icon is created

joassouza committed:
[chore] Clean up

lpagliari commented on static/js/commentSaveOrDelete.js:
```diff
@@ -2,7 +2,7 @@ var _ = require('ep_etherpad-lite/static/js/underscore');
 exports.saveCommentOnPreMarkedText = function(commentId, preMarkedTextRepArr, ace) {
-  var attributeName = 'comment';
+  var attributeName = commentId;
   var attributeValue = commentId;
```
Algum motivo para não usar apenas `true`/`false` como valor? Não cheguei a olhar o código todo, imagino que isso possa ser útil em outro lugar...

joassouza replied:
não tive motivo, deixei o valor do jeito que estava. Pensando bem, não faz sentido mesmo deixar o commentID como o valor do atributo. Vou mudar

lpagliari replied:
antes de vc fazer a mudança para `true`/`false`, talvez seja bom ver a coisa do nome do atributo ser mais do que apenas `c-...`. Se vc usar um prefixo, por ex, dá para deixar o `commentId` no valor mesmo, assim não precisa fazer parse do nome do atributo quando precisar pegar o `commentId` de novo.

lpagliari commented on static/js/index.js:
```diff
@@ -417,7 +417,7 @@ var hooks = {
   // Insert comments classes
   aceAttribsToClasses: function(hook, context){
-    if(context.key === 'comment' && context.value !== "comment-deleted") {
+    if(context.key.startsWith(shared.COMMENT_PREFIX) && context.value !== "comment-deleted") {
```
Fiquei um pouco com medo de estarmos usando apenas `c-` como prefixo aqui -- medo de dar conflito com outro atributo. Acha muito arriscado?

joassouza replied:
vou colocar uma condição mais forte no if, tipo usar uma regex para validar se é o atributo

joassouza replied:
um conflito de atributo seria um bug muito, muito tenso de encontrar se acontecesse. Melhor adotar um approach mais defensivo mesmo

lpagliari commented on static/tests/frontend/specs/api-dataChanged.js:
```diff
@@ -196,6 +203,11 @@ describe('ep_comments_page - api - "data changed" event', function() {
+      after(function () {
+        // clean the comments created, and the html inserted
+        _.times(6, utils.undo);
+      });
```
Fiquei pensando aqui sobre esses testes... Embora não tenha um arquivo melhor para colocá-los (dentre os existentes), eu achei que eles aqui não tinham muito a ver com a parte da API que manda updates dos dados. Quem sabe cria um arquivo específico para os testes de dois comentários no mesmo trecho, que tal?

Esse `_.times(6, utils.undo);`, por exemplo, acabou sendo necessário pq estamos tentando aproveitar testes que não tem muito a ver com o que a gente está testando aqui.

joassouza replied:
Eu concordo viu. Eu tive o mesmo feeling quando comecei a _procurar lugar_ para colocar os testes dessa feature

lpagliari commented on static/tests/frontend/specs/api-dataChanged.js:
```diff
@@ -283,5 +299,30 @@ describe('ep_comments_page - api - "data changed" event', function() {
+
+      it('sends the data of created comment', function(done) {
+        var comments = apiUtils.getLastDataSent();
+        expect(comments.length).to.be(2);
```
Acho importante testar os dados mandados, mesmo que só uma parte deles (o texto do comentário, por ex). Assim evitamos bugs como estarmos mandando um comentário deletado, ou 2x o mesmo comentário.

lpagliari commented on static/tests/frontend/specs/commentIcons.js:
```diff
@@ -398,5 +398,24 @@ describe('ep_comments_page - Comment icons', function() {
-  })
+  });
+
+  // here we create a new clean pad to reset the previous state that simulates two pages
```
2 páginas?

joassouza replied:
tá errado. Era para _dois usuários editando o mesmo script_

lpagliari commented on static/tests/frontend/specs/commentIcons.js:
```diff
@@ -398,5 +398,24 @@ describe('ep_comments_page - Comment icons', function() {
+    before(function(done) {
+      var test = this;
+      createScript(test, function(){
+        // we create a comment so we create a second one that wraps the first one
```
Boa!! 👍

lpagliari commented on static/js/commentDataManager.js:
```diff
@@ -166,65 +167,86 @@ commentDataManager.prototype.triggerDataChanged = function() {
-      commentId: commentId,
-      element: this,
-    };
+  $commentsOnText.each(function() {
```
não dá para usar `map` ao invés de `each`? É mais direto...

joassouza replied:
acho que usei o `each` por que poderia ter mais de um comentário por trecho o que BTW  não faz nenhum sentido 😒 . O próprio `map` daria conta já que faço o `flatten` depois de executar esse trecho

lpagliari commented on static/js/commentDataManager.js:
```diff
@@ -166,65 +167,86 @@ commentDataManager.prototype.triggerDataChanged = function() {
-    commentData.replies = _(commentData.replies).pick(sortedReplyIds);
+commentDataManager.prototype._getSceneNumber = function ($scenes, nodeWithComment) {
+  // fill scene number
+  // go to a function maybe we call it on _getListOfCommentsOrdered
```
não entendi o comentário dessa linha aqui...

joassouza replied:
era para remover isso, era parte de umas anotações que eu tinha feito

lpagliari commented on static/js/commentDataManager.js:
```diff
@@ -166,65 +167,86 @@ commentDataManager.prototype.triggerDataChanged = function() {
-      return commentData.replies[replyId].timestamp;
-    });
-    commentData.replies = _(commentData.replies).pick(sortedReplyIds);
+commentDataManager.prototype._getSceneNumber = function ($scenes, nodeWithComment) {
```
ficou bem melhor extrair para essa função! 👍

lpagliari commented on static/js/commentDataManager.js:
```diff
@@ -166,65 +167,86 @@ commentDataManager.prototype.triggerDataChanged = function() {
+    var isAReplyId = shared.getReplyIdsFrom(className).length;
+
+    // there might be another comment (with reply) on the same DOM nodeWithComment of this comment
+    var isAReplyIdOfThisComment = isAReplyId && commentData.replies[className];
```
Talvez seja uma boa fazer o teste com comentários + replies, assim testa essa parte tb (se eu entendi direito o que essa parte faz, hehe)

joassouza replied:
essa parte não é muito claro mesmo. Eu removi esse código de uma função e coloquei aí. Na verdade, meus planos é mudar um pouco essa lógica. Se eu conseguir fazer aquela otimização que nós falamos se brincar nem precisará disso

joassouza committed:
[refactor:test] Extract tests to a separated file

Move the tests of adding two comments on a same range of text to a
separated file

joassouza committed:
[refactor] Change comment attribute key
